File: Semaphore_sharp/Semaphore_sharp/Storage.cs
using System.Collections.Generic;
using System.Threading;

namespace Semaphore_sharp
{
    public class Storage
    {
        public Semaphore AccessToStorageSemaphore;
        public Semaphore PushItemSemaphore;
        public Semaphore PickItemSemaphore;

        private List<string> _storage;

        public Storage(int storageSize)
        {
            AccessToStorageSemaphore = new Semaphore(1 ,1);
            PushItemSemaphore = new Semaphore(storageSize, storageSize);
            PickItemSemaphore = new Semaphore(0, storageSize);

            _storage = new List<string>(storageSize);
        }

        public void PushItem(string itemName)
        {
            _storage.Add(itemName);
        }

        public string PickItem()
        {
            string itemName = _storage[0];
            _storage.RemoveAt(0);
            return itemName;
        }
    }
}

File: Semaphore_java/src/Storage.java
import java.util.LinkedList;
import java.util.concurrent.Semaphore;

public class Storage {
    public final Semaphore AccessToStorageSemaphore;
    public final Semaphore PushItemSemaphore;
    public final Semaphore PickItemSemaphore;

    private LinkedList<String> _storage;

    public Storage(int storageSize){
        AccessToStorageSemaphore = new Semaphore(1);
        PushItemSemaphore = new Semaphore(storageSize);
        PickItemSemaphore = new Semaphore(0);

        _storage = new LinkedList<>();
    }

    public void pushItem(String itemName){
        _storage.add(itemName);
    }

    public String pickItem(){
        String itemName = _storage.get(0);
        _storage.removeFirst();
        return itemName;
    }
}